tiatoolbox/dataloader/wsireader.py

```diff
@@ -82,39 +91,524 @@ def slide_info(self):
+                power, level, baseline. Baseline refers to the largest
+                resolution in the WSI (level 0).
+
+        Raises:
+            ValueError: Missing MPP metadata
+            ValueError: Missing objective power metadata
+            ValueError: Invalid units
+
+        Returns:
+            list: Scale for each level relative to the given scale and
+                units
+
+        Examples:
+            >>> from tiatoolbox.dataloader import wsireader
+            >>> wsi = wsireader.WSIReader("CMU-1.ndpi")
+            >>> print(wsi._relative_level_scales(0.5, "mpp"))
+            [array([0.91282519, 0.91012514]), array([1.82565039, 1.82025028]) ...
+
+            >>> from tiatoolbox.dataloader import wsireader
+            >>> wsi = wsireader.WSIReader("CMU-1.ndpi")
+            >>> print(wsi._relative_level_scales(0.5, "baseline"))
+            [0.125, 0.25, 0.5, 1.0, 2.0, 4.0, 8.0, 16.0, 32.0]
+        """
+        info = self.slide_info
+
+        def make_into_array(x):
+            """Ensure input x is a numpy array of length 2."""
+            if isinstance(x, numbers.Number):
+                # If one number is given, the same value is used for x and y
+                return np.array([x] * 2)
+            return np.array(x)
+
+        @np.vectorize
+        def level_to_downsample(x):
+            """Get the downsample for a level, interpolating non-integer levels."""
+            if isinstance(x, int) or int(x) == x:
+                # Return the downsample for the level
+                return info.level_downsamples[int(x)]
+            # Linearly interpolate between levels
+            floor = int(np.floor(x))
+            ceil = int(np.ceil(x))
+            floor_downsample = info.level_downsamples[floor]
+            ceil_downsample = info.level_downsamples[ceil]
+            return np.interp(x, [floor, ceil], [floor_downsample, ceil_downsample])
+
+        resolution = make_into_array(resolution)
+
+        if units == "mpp":
+            if info.mpp is None:
+                raise ValueError("MPP is None")
+            base_scale = info.mpp
+        elif units == "power":
+            if info.objective_power is None:
+                raise ValueError("Objective power is None")
+            base_scale = 1 / info.objective_power
+            resolution = 1 / resolution
+        elif units == "level":
+            if any(resolution >= len(info.level_downsamples)):
+                raise ValueError("Target scale level > number of levels in WSI")
+            base_scale = 1
+            resolution = level_to_downsample(resolution)
+        elif units == "baseline":
+            base_scale = 1
+            resolution = 1 / resolution
+        else:
+            raise ValueError("Invalid units")
+
+        return [(base_scale * ds) / resolution for ds in info.level_downsamples]
+
+    def _find_optimal_level_and_downsample(self, resolution, units, precision=3):
+        """
```

Comment:
Text in the docstring should start at this line.

Reply:
Dome in https://github.com/TIA-Lab/tiatoolbox/pull/40/commits/e80b7d5e78c98de99502d3f49b32da784de0ac9d